#160 Remove complex widget

MPParsley opened the pull request.

amitaibu committed:
Start removing complex widget

amitaibu committed:
Add core widgets

amitaibu committed:
Allow privileged user to create content

amitaibu committed:
Check permissions

amitaibu committed:
Start fixing tests

amitaibu committed:
Fix logic and tests

amitaibu committed:
Sniffer fixes

amitaibu committed:
Fix return value in docs

amitaibu committed:
Re-add Functional test

amitaibu committed:
Sniffer fixes

amitaibu committed:
Sniffer fixes

amitaibu committed:
Merge branch '8.x-1.x' into 160-remove-complex-widget

Commit:
Setting up the base for the new test.

Commit:
Commit more work.

Commit:
First version of the tests.

Commit:
Make code sniffer happy.

amitaibu committed:
Merge 8.x-1.x

amitaibu committed:
Remove un-needed method

amitaibu committed:
Remove unused statement

amitaibu committed:
Start expanding the widget tests

amitaibu committed:
Create node types

amitaibu committed:
Comments cleanup

amitaibu committed:
Remove traits

amitaibu committed:
Try to find cause of bug

amitaibu committed:
Remove unusued code

amitaibu committed:
Remove unusued code

amitaibu committed:
Remove custom settings

amitaibu committed:
Remove custom settings

amitaibu committed:
Remove unused use

amitaibu committed:
Add todo

amitaibu committed:
Filter out the bundles that are not groups.

amitaibu committed:
Filter out the bundles that are not groups.

amitaibu committed:
Fix indentation

amitaibu committed:
Fix indentation

amitaibu committed:
Improve exception error

amitaibu committed:
Improve names

amitaibu committed:
Sniffer fixes

amitaibu committed:
Convert test to Functional

amitaibu committed:
Add comment

amitaibu committed:
Debug failing tests

amitaibu committed:
Revert debug

amitaibu committed:
Fix wrong node type name

amitaibu committed:
Simplify and debug test

amitaibu committed:
Add status 200

amitaibu committed:
gst

amitaibu committed:
Quicker debugs

amitaibu committed:
Check 404 page

amitaibu committed:
Rebuild routes

amitaibu committed:
Fields page

amitaibu committed:
click link

amitaibu committed:
Back to drupalGet

amitaibu committed:
Check access

amitaibu committed:
Enable field ui

amitaibu committed:
Allow debug

amitaibu committed:
Import URL

amitaibu committed:
Debug user access

amitaibu committed:
Cleanup

amitaibu committed:
Admin user

amitaibu committed:
Back to permissions

amitaibu committed:
Remove use

amitaibu committed:
Permissions

amitaibu committed:
Core test

amitaibu committed:
Run core

amitaibu committed:
Debug core

amitaibu committed:
Remove core debug

amitaibu committed:
Start conveting test to Kernel

amitaibu committed:
More work

amitaibu committed:
Use entity test with bundle

amitaibu committed:
Merge field settings

amitaibu committed:
Fix test

amitaibu committed:
Sniffer fixes

amitaibu committed:
Remove debug

amitaibu committed:
Create Node type correctly

amitaibu committed:
Add unpublished node

amitaibu committed:
Fix OgSelectionWidgetTest

Commit:
Validate that the user have access to post a group content.

Commit:
Adding skeleton for a test [skip ci]

Commit:
Commit work [skip ci]

Commit:
Fix a typo and continue working on the test [skip ci]

Commit:
Set the role properly [skip ci]

Commit:
Use forbidden and not allowed.

Commit:
Checkin when forbidden.

Commit:
Somehow fix the tests.

Commit:
Getting the root entity.

Commit:
Merge branch '8.x-1.x' into 160-remove-complex-widget

Commit:
Merge branch '8.x-1.x' into 160-remove-complex-widget

Commit:
Fix conflict

Commit:
Stbalize tests.

Commit:
Minor refactor.

Commit:
Wrong operator.

Commit:
Trying to understand what’s the difference [skip ci]

Commit:
Find another solution.

Commit:
Adding basic constraint to add entities [skip ci]

Commit:
Commit logic for now.

Commit:
Don’t add constant to all the entities, just check the field.

Commit:
Fix tests.

Commit:
Remove the testin file

Commit:
Fix name.

Commit:
Fix cs

Commit:
Get the text.

Commit:
Maybe now?

Commit:
Catching and not version.

Commit:
CS fix.

Commit:
Handle another case for the fields.

Commit:
Set the field as required in specific cases.

Commit:
More changes.

Commit:
Adding the validation.

Commit:
The validation works as expected. Now, we need to write tests.

Commit:
More improvements.

Commit:
Adding the boilerplate for the new test.

Commit:
Done with the tests.

Commit:
Minor fixes.

Commit:
Make CS happy.

Commit:
make cs happy.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into 160-remove-complex-widget

mariano-dagostino committed:
Fix grammar in comments.

mariano-dagostino committed:
Improve comment of access check logic

mariano-dagostino committed:
GroupTypeManager::getAllGroupBundles should respect GroupTypeManagerInterface method signatures.

RoySegall committed:
Merge branch '8.x-1.x' into 160-remove-complex-widget

MPParsley committed:
Use single quotes

MPParsley committed:
Undo single quotes

MPParsley committed:
Improve readability foreach

MPParsley committed:
Updated documentation for the entity reference field

MPParsley committed:
Use invalidMessage from constraint

MPParsley committed:
Added dependency injection to ValidOgMembershipMultipleReferenceConstraintValidator

pfrenssen committed:
Merge remote-tracking branch 'gizra/8.x-1.x' into 160-remove-complex-widget

MPParsley committed:
Fix tests

MPParsley committed:
Fixed unnecessarily gendered language in a comment

MPParsley committed:
Cleanup call to MembershipManager::reset()

Removed in #480.

MPParsley committed:
Protected properties should start with a small cap

MPParsley committed:
Protected properties should start with a small cap.

MPParsley committed:
Update src/Plugin/Validation/Constraint/ValidOgMembershipMultipleReferenceConstraintValidator.php

MPParsley committed:
Revert commit

MPParsley committed:
Deprecate Drupal\simpletest\ContentTypeCreationTrait

See https://www.drupal.org/node/2884454

MPParsley committed:
Deprecated Drupal\simpletest\NodeCreationTrait

See https://www.drupal.org/node/2884454

MPParsley committed:
Fixed namespaces

MPParsley committed:
Removed deprecated call to Drupal::entityManager

MPParsley committed:
Removed deprecated entity.query service

MPParsley committed:
Removed calls to deprecated entityManager

MPParsley committed:
Use EntityTypeManagerInterface

MPParsley committed:
Remove call to deprecated entity.manager

MPParsley committed:
Fixed EntityManager deprecations in OgSelection

See https://www.drupal.org/node/2549139

MPParsley committed:
Fixed call to undefined function getBundleInfo() on EntityTypeManager

MPParsley committed:
Added getQuery

MPParsley committed:
Clear membership cache

MPParsley committed:
Restore cache reset

MPParsley committed:
Test with cache delete.

MPParsley committed:
Cleanup test

MPParsley committed:
Fixed indentation

MPParsley committed:
Added missing use statement

MPParsley committed:
Fix tests

MPParsley committed:
Revert permission format

MPParsley committed:
Revert permission format

MPParsley committed:
Revert permission format

damienmckenna committed:
Merge branch '8.x-1.x' into 160-remove-complex-widget

MPParsley committed:
Merge pull request #569 from damienmckenna/160-remove-complex-widget

160 remove complex widget

MPParsley committed:
Remove unused use statement

damienmckenna committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

pfrenssen committed:
Merge pull request #595 from damienmckenna/feature/160-remove-complex-widget

Fix merge conflicts from 8.x-1.x

MPParsley committed:
Fix codesniffer errors

use lowerCamel naming for class properties

damienmckenna committed:
Removed redundant getGroupBundlesByEntityType() method.

damienmckenna committed:
Typoo.

MPParsley committed:
Merge pull request #596 from damienmckenna/feature/570-getGroupBundlesByEntityType

Removed redundant getGroupBundlesByEntityType() method.

MPParsley committed:
Revert "Removed redundant getGroupBundlesByEntityType() method."

MPParsley committed:
Merge pull request #597 from Gizra/revert-596-feature/570-getGroupBundlesByEntityType

Revert "Removed redundant getGroupBundlesByEntityType() method."

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Fix merge conflict

MPParsley committed:
Removed unused variables

MPParsley committed:
Update og.module

MPParsley committed:
Test without admin field_mode

MPParsley committed:
Test with deprecated Og::getSelectionHandler

MPParsley committed:
fix merge conflict

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge pull request #636 from Gizra/8.x-1.x

Merge develop into 570

MPParsley committed:
undo remove whitespace

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

# Conflicts:
#	src/Plugin/Field/FieldWidget/OgComplex.php

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

# Conflicts:
#	src/Plugin/Field/FieldWidget/OgComplex.php
#	tests/src/Functional/OgSelectionWidgetTest.php

MPParsley committed:
Fix deprecation

MPParsley committed:
Renamed permission

#661 Rename the global permission to administer all groups to 'administer organic groups'

MPParsley committed:
Remove deprecated method reset()

MPParsley committed:
Deprecate handler_settings

Providing settings under 'handler_settings' is deprecated in drupal:8.4.0 support for 'handler_settings' is removed from drupal:9.0.0. Move the settings in the root of the configuration array. See https://www.drupal.org/node/2870971

MPParsley committed:
Remove deprecated call to Og::getSelectionHandler

MPParsley committed:
Fix undefined index target_bundles

MPParsley committed:
Remove deprecated call to Og::getSelectionHandler

MPParsley committed:
Revert Og::getSelectionHandler

MPParsley committed:
Apply suggestions from code review

Co-authored-by: Damien McKenna <damien@mc-kenna.com>

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Split up OgAccess::userAccessEntity

See #673.

MPParsley committed:
Fix order of parameters

MPParsley committed:
Fix syntax error

MPParsley committed:
Added period.

MPParsley committed:
Inject services

- Drupal\Core\Session\AccountInterface
- Drupal\og\OgAccessInterface

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Revert injection of OgAccess

MPParsley committed:
Revert injection of current user

MPParsley committed:
Remove hack for Drupal 8.4

MPParsley committed:
Revert dependency injections

MPParsley committed:
Revert hack for ValidReferenceConstraintValidator

Commit:
Inject services.

Commit:
Fix tests

MPParsley committed:
Test deprecation of Og:getSelectionHandler

MPParsley committed:
Use existing method getGroupBundleIdsByEntityType

MPParsley committed:
Revert getAllGroupBundles

MPParsley committed:
Remove getGroupBundlesByEntityType

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

# Conflicts:
#	src/Plugin/Field/FieldWidget/OgComplex.php

MPParsley committed:
Revert deprecation

MPParsley committed:
Revert deprecation

MPParsley committed:
Fix strict types

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x'

Resolve merge conflicts.

MPParsley committed:
Apply suggestions from code review

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Fix merge conflict

MPParsley commented on tests/src/Kernel/EntityReference/OgSelectionConfigurationFormTest.php:
```diff
@@ -0,0 +1,98 @@
+    'field',
+    'field_ui',
+    'entity_test',
+    'og',
```
```suggestion
    'og',
    'options',
```

MPParsley committed:
Add dependency to options

Fix PluginNotFoundException: The "field_item:list_string" plugin does not exist.

damienmckenna commented:
Is this intended to be a cleaned-up copy of the original PR? I've lost track of what's what.

MPParsley commented:
@damienmckenna, indeed. I had closed #570 and fixed some merge conflicts and a broken test in this PR.

Perhaps you can have a look at the deprecation error?

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

zero2one commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -81,37 +155,45 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
-      foreach ($user_groups as $group) {
+    foreach ($this->getUserGroups() as $group) {
+      // Check user has "create" permission on this entity.
+      $permission = "create $bundle content";
```
This should be:
```php
$permission = sprintf('create %s %s', $bundle, $entity->getEntityTypeId());
```

MPParsley commented on src/Plugin/EntityReferenceSelection/OgSelection.php:
```diff
@@ -81,37 +155,45 @@ protected function buildEntityQuery($match = NULL, $match_operator = 'CONTAINS')
-      // Don't include the groups, the user doesn't have create permission.
-      foreach ($user_groups as $group) {
+    foreach ($this->getUserGroups() as $group) {
+      // Check user has "create" permission on this entity.
+      $permission = "create $bundle content";
```
@zero2one, we should use the permission layer instead of using string patterns:
```suggestion
      // Check if the current user can perform the "create" operation on this entity.
      if ($this->ogAccess->userAccessGroupContentEntityOperation('create', $group, $entity, $this->currentUser)) {
```

zero2one replied:
Then we need to check if we have a new or existing entity ($entity->isNew()) and do something like this:

* Entity is new => check "create" operation for that group.
* Entity is existing and linked to the group being checked: check "update" operation for that group.
* Entity is existing and not linked to the group being checked : check "create" operation for that group.

pfrenssen commented:
I started reviewing this, but right at the start of the code review I see the same bizarre and incorrect logic that I already reported back in 2019 on #570.

Can we try not to continue on this flawed approach? By continuing on this path and not addressing the fundamental flaws in it we are just wasting time. Maybe if the approach is generally considered to be correct and there are only some rough edges that I happen to stumble over, would it be possible to at least remove all code changes that are either undocumented, untested or have any other bizarre intentions / code smells?

pfrenssen commented on og.module:
```diff
@@ -164,6 +164,20 @@ function og_entity_access(EntityInterface $entity, $operation, AccountInterface
   /** @var \Drupal\Core\Access\AccessResult $access */
   $access = \Drupal::service('og.access')->userAccessEntityOperation($operation, $entity, $account);
 
+  $audience_fields = \Drupal::service('og.group_audience_helper')->getAllGroupAudienceFields($entity_type_id, $bundle_id);
+  if (count($audience_fields) === 1) {
+    $field_definitions = \Drupal::service('entity_field.manager')->getFieldDefinitions($entity_type_id, $bundle_id);
+    foreach ($field_definitions as $field_definition) {
+      /** @var \Drupal\Core\Field\FieldDefinitionInterface $field_definition */
+      if (!in_array($field_definition->getName(), array_keys($audience_fields))) {
+        continue;
+      }
+
+      // Check if the field needs to be required.
+      $field_definition->setRequired($access);
+    }
+  }
+
```
This is logic is undocumented, untested and seemingly unneeded. Altering field definitions to set them required inside `hook_entity_access()`? This is completely bizarre and can not be the correct solution for whatever this is trying to fix.

This section MUST be removed.

pfrenssen commented on og.module:
```diff
@@ -201,6 +215,11 @@ function og_og_membership_type_access(OgMembershipTypeInterface $entity, $operat
 function og_entity_create_access(AccountInterface $account, array $context, $bundle) {
   $entity_type_id = $context['entity_type_id'];
 
+  if (!empty($context['skip_og_permission'])) {
```
This workaround `skip_og_permission` indicates a flaw in the logic of this PR. We should not introduce a bypass in access checking. This will just lead to more complexity down the road and potential security vulnerabilities. There must be a loop introduced somewhere, where an access check is done inside of an access check. We should fix this in a different way.

pfrenssen commented on og.module:
```diff
@@ -226,23 +245,41 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
       continue;
     }
 
+    if (count($audience_fields) === 1) {
+      // Set the field to required only if we have a single audience field.
+      // If we have multiple fields then the require logic will perform at the
+      // entity constraint level.
+      $field_definition->setRequired(!\Drupal::entityTypeManager()
+        ->getAccessControlHandler($entity_type_id)
+        ->createAccess($bundle, $account, ['skip_og_permission' => TRUE]));
+    }
```
What is going on here? Again altering field definitions inside access hooks???

This change MUST be removed. This is NOT the place for messing with anything else that is not strictly about access.

pfrenssen commented on og.module:
```diff
@@ -226,23 +245,41 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
     $handler = \Drupal::service('plugin.manager.entity_reference_selection');
+    $handler = $handler->getInstance($options);
+
+    // @todo Remove deprecated call to Og::getSelectionHandler.
+    $handler = Og::getSelectionHandler($field_definition);
```
Introducing deprecated calls in new code?

pfrenssen commented on og.module:
```diff
@@ -226,23 +245,41 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
 
-    if ($handler->getInstance($options)) {
-      return AccessResult::neutral();
+    if ($handler && $handler->countReferenceableEntities()) {
+      // At least one of the fields has referenceable groups. That mean that
+      // the user can create group content when referencing the group content in
+      // at least one of the fields.
+      return AccessResult::allowed();
```
This is not clear to me, but it raises a red flag. An additional check is done, but why is the access level raised from neutral to allowed?

pfrenssen reviewed (changes requested):
I have reviewed 10% of the PR. There are some serious problems with what I have seen so far.

pfrenssen commented on og.module:
```diff
@@ -226,23 +245,41 @@ function og_entity_create_access(AccountInterface $account, array $context, $bun
 
+  /** @var \Drupal\og\OgGroupAudienceHelperInterface $audience_helper */
+  $audience_helper = \Drupal::service('og.group_audience_helper');
+  $audience_fields = $audience_helper->getAllGroupAudienceFields($entity_type_id, $bundle);
   $field_definitions = \Drupal::service('entity_field.manager')->getFieldDefinitions($entity_type_id, $bundle);
```
Both these calls to `::getAllGroupAudienceField()` and `::getFieldDefinitions()` duplicate the same actions. The field definition list is retrieved twice and filtered twice now.

amitaibu commented:
@pfrenssen after so many years that this PR is laying around, it's possible that what's best is to close it, and start from scratch (if you have the will/ capacity for it :wink: ).

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley commented on tests/src/Functional/OgMultipleSelectionWidgetAutoCompleteTest.php:
```diff
@@ -0,0 +1,199 @@
+        'create group_content content',
+        'edit own group_content content',
+        'edit any group_content content',
+      ],
```
Drupal\Tests\og\Functional\OgMultipleSelectionWidgetAutoCompleteTest::testAutoCompleteForNonGroupMember
Role label is expected to be a string. in /home/travis/build/Gizra/drupal/core/modules/user/user.module:851

```suggestion
      ],
      'label' => 'Dummy role',
```

MPParsley committed:
Add role label

MPParsley committed:
Merge 8.x

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

# Conflicts:
#	tests/src/Functional/OgSelectionWidgetTest.php

MPParsley committed:
Merge branch '8.x-1.x' into feature/160-remove-complex-widget

MPParsley committed:
Fix fatal error

Declaration of Drupal\Tests\og\Functional\OgMultipleSelectionWidgetAutoCompleteTest::setUp() must be compatible with Drupal\Tests\BrowserTestBase::setUp(): void in /home/travis/build/Gizra/og/tests/src/Functional/OgMultipleSelectionWidgetAutoCompleteTest.php on line 65

MPParsley committed:
Fix fatal error

Declaration of setUp() must be compatible with Drupal\Tests\BrowserTestBase::setUp(): void

MPParsley committed:
Fix fatal error

Declaration of setUp() must be compatible with Drupal\Tests\BrowserTestBase::setUp(): void

MPParsley committed:
Fix deprecation notices

The Drupal\Tests\og\Kernel\EntityReference\OgSelectionConfigurationFormTest::$modules property must be declared protected. See https://www.drupal.org/node/2909426

MPParsley committed:
Fix deprecation notices

The Drupal\Tests\og\Kernel\EntityReference\OgSelectionConfigurationFormTest::$modules property must be declared protected. See https://www.drupal.org/node/2909426

claudiu-cristea commented:
Since https://www.drupal.org/project/og/issues/3467694 was merged into the `8.x-1.x` branch, the development has been moved to https://www.drupal.org/project/og. If this change is still actual, please open an issue at https://www.drupal.org/project/issues/og and provide there a _merge request_ similar with this pull request.

Thank you for contributing to Organic groups module.
